src/renderers/shared/reconciler/ReactReconciler.js

```diff
@@ -66,7 +66,9 @@ var ReactReconciler = {
   ) {
     var prevElement = internalInstance._currentElement;
 
-    if (nextElement === prevElement && nextElement._owner != null) {
+    if (nextElement === prevElement &&
+        nextElement._owner != null &&
+        Object.keys(context).length === 0) {
```

Comment:
We have the old context stored if it is a composite so we should just be able to do an object comparison. This check is too slow to do for every hot path, but more importantly it means that you never get this optimization if you use context anywhere in your tree. Which is like everywhere.